src/hotspot/share/utilities/istream.hpp

```diff
@@ -0,0 +1,449 @@
+// Use the sizing access functions if you care about this.
+//
+// Formatting guidelines:
+//
+// Configuration data should be line-oriented.  It should be readable
+// by humans (though perhaps with difficulty).  It should be easily
+// processed by text editors and by widely available text processing
+// tools such as grep, sed, and awk.
+//
+// Configuration data should not require "compilers" to generate, if
+// possible.  It should be editable by hand, if possible.  In cases
+// where binary data is strongly required, pick a binary format
+// already native to Hotspot, such as classfile, jar, or jmod.
+//
+// Each line should be separately parseable; the parsing can be ad
+// hoc.  For constructs inherently larger than single lines (such as
+// complex method configuration information), try to use a structuring
+// principle that allows "leaf" data to be line-oriented, and delimits
+// that data with markup lines of some sort.  Try to pick a
+// line-friendly version of a standard format like XML or Markdown.
+// JSON is somewhat problematic because there is no line-friendly leaf
+// syntax: everything at the leaves must be a quoted string in JSON.
+//
+// Use simple parsing via scanf-like formats for simple applications.
+// But, keep in mind that these formats may lose data when applied to
+// unusual strings, such as class names that contain spaces, or method
+// names that contain punctuation.  For more robust transmission of
+// potentially unusual names, consider wrapping them in XML-flavored
+// lines like <tag attr='pay load'/>.
+//
+// Note: Input streams are never MT-safe.
+
+class inputStream : public CHeapObjBase {
+ public:
+  class Input;
+
+ private:
+  NONCOPYABLE(inputStream);
+
+  static constexpr size_t SMALL_SIZE =  240 DEBUG_ONLY(*0 + 10);
+  static constexpr size_t BIG_SIZE   = 2048 DEBUG_ONLY(*0 + 20);
+
+ protected:
+  // Values for _input_state, to distinguish some phases of history:
+  // Do we need to read more input (NTR)?  Did we see EOF already?
+  // Was there an error getting input or allocating buffer space?
+  enum class IState : int { NTR_STATE, EOF_STATE, ERR_STATE };
+
+  // Named offset for _next relative to _content_end, of phantom '\n'.
+  static const int NEXT_PHANTOM = 1;
+
+  Input* _input;   // where the input comes from or else nullptr
+  IState _input_state;  // one of {NTR,EOF,ERR}_STATE
+  char   _line_ending;  // one of {0,1,2} for "", "\n", "\r\n"
+  char*  _buffer;       // scratch buffer holding at least the current line
+  size_t _buffer_size;  // allocated size of buffer
+  size_t _content_end;  // offset to end of valid contents of buffer
+  size_t _beg;          // offset in buffer to start of current line
+  size_t _end;          // offset to end of known current line (else content_end)
+  size_t _next;         // offset to known start of next line (else =end)
+  void*  _must_free;    // unless null, a malloc pointer which we must free
```

Comment:
Reading this code, why do we set `_must_free` instead of simply having a method:

```c++
bool must_free() {
  return _buffer != &_small_buffer;
}
```

and just delete the `_must_free` field.

Reply:
Good question.  There was a version of the code that accepted a user-supplied buffer, optionally.  In that case, `_must_free` was set false (or to a user-requested value), since it was up to the user whether the user-supplied buffer should be freed.  It could be a static buffer.  If there is no longer such an option in the existing constructors, then this field can be GC-ed.

Reply:
I removed `_must_free` and added a new `has_c_heap_buffer()` method.